gk/main.c

```diff
@@ -968,12 +974,215 @@ gk_setup_rss(struct gk_config *gk_conf)
 	return ret;
 }
 
+static inline unsigned short
+icmp_cksum(void *buf, unsigned int size)
+{
+	unsigned short *buffer = buf;
+	unsigned long cksum = 0;
+
+	while(size > 1) {
+		cksum += *buffer++;
+		size -= sizeof(*buffer);
+	}
+
+	if(size)
+		cksum += *(unsigned char*)buffer;
+
+	cksum = (cksum >> 16) + (cksum & 0xffff);
+	cksum += (cksum >> 16);
+
+	return (unsigned short)(~cksum);
+}
+
+static void
+xmit_icmp(struct gatekeeper_if *iface, struct ipacket *packet,
+	uint16_t *num_pkts, struct rte_mbuf **icmp_bufs)
+{
+	struct ether_addr eth_addr_tmp;
+	struct ether_hdr *icmp_eth;
+	struct ipv4_hdr *icmp_ipv4;
+	struct icmp_hdr *icmph;
+	struct rte_mbuf *pkt = packet->pkt;
+	int icmp_pkt_len = iface->l2_len_out + sizeof(struct ipv4_hdr) +
+		sizeof(struct icmp_hdr);
+	if (pkt->data_len > icmp_pkt_len) {
+		int ret = rte_pktmbuf_trim(pkt, pkt->data_len - icmp_pkt_len);
+		if (ret < 0) {
+			RTE_LOG(ERR, MBUF,
+				"Failed to remove %d bytes of data at the end of the mbuf at %s.",
+				pkt->data_len - icmp_pkt_len, __func__);
+			drop_packet(pkt);
+			return;
+		}
+
+		icmp_eth = rte_pktmbuf_mtod(pkt, struct ether_hdr *);
+	} else {
+		icmp_eth = (struct ether_hdr *)rte_pktmbuf_append(pkt,
+			icmp_pkt_len - pkt->data_len);
+		if (icmp_eth == NULL) {
+			RTE_LOG(ERR, MBUF,
+				"Failed to append %d bytes of new data: not enough headroom space in the first segment at %s!\n",
+				icmp_pkt_len - pkt->data_len, __func__);
+			drop_packet(pkt);
+			return;
+		}
+	}
+
+	eth_addr_tmp = icmp_eth->s_addr;
+	ether_addr_copy(&icmp_eth->d_addr, &icmp_eth->s_addr);
```

Comment:
Why did you avoid using `ether_addr_copy()` at the line `eth_addr_tmp = icmp_eth->s_addr;` as in the following lines?  See analogous condition in `xmit_icmpv6()`.